### ISSUE TITLE
cy.select() won't work, need to force and select twice

### ISSUE THREAD

**Issue by unfadding** [cypress-io/cypress#4335]
### Current behavior:

```js
cy.get('#element').select(value) //won't work

cy.get('#element').select(value, { force: true });
cy.get('#element').select(value, { force: true }); //only works calling twice, using force
```

### Desired behavior:

```
cy.get('#element').select(value) //want it to work like this
```

### Versions

Cypress 3.3.1
Chrome 74
Windows 10 x64

**Comment by jennifer-shehane** [comment]
Hey @unfadding, we have several tests that this works fine in Cypress. 

Can you provide a reproducible example that we can run locally to see this behavior? It likely has something to do with the html, css, or js on your particular website, so providing a small repro is the only way we can address this - otherwise we'll have to close the issue.

**Comment by jennifer-shehane** [comment]
Unfortunately we have to close this issue as there is not enough information to reproduce the problem. This does not mean that your issue is not happening - it just means that we do not have a path to move forward. 

Please comment in this issue with a reproducible example and we will reopen the issue. 🙏

**Comment by aasthasanghi** [comment]
Hi @jennifer-shehane,
I'm also facing the same issue in cypress 8.3.0. 

Here is a snippet using which you can reproduce the issue:

```
cy.visit('https://apply.kiwibank.co.nz/credit-cards/low-rate-visa/my-personal-details');
cy.get('#timeAtAddress').select('0 - 6 months');
```
here drop-down is not selected on screen until I run line 2 twice. Cypress doesn't show any error but the drop-down is also not selected on the browser.

I'm using 
Windows 10
Cypress 8.3.0

Please advise.

Can you

**Comment by OliJeffery** [comment]
I'm still getting this occasionally in Cypress 12

**Comment by Lukas-R4** [comment]
works for me when i use interpolation

``cy.get('#element').should('be.visible').select(`${value}`)``

**Comment by wendyn-projects** [comment]
I have the same issue
Cypress 12.5.1
Windows 10 Enterprise _19044.2604_
Chrome 110 _.0.5481.105_

the code I tried:
```js
cy.xpath("//select[@data-XXX='XXX']").first({ log: false }).select(value);
cy.xpath("//select[@data-XXX='XXX']").select(value);
```
I also tried
```js
cy.xpath("//select[@data-XXX='XXX']").first({ log: false }).select(value).should("have.text", value);
```
which strangely enough **worked** but obviously I got AssertionError,
however
```js
cy.xpath("//select[@data-XXX='XXX']").first({ log: false }).select(value).find(':selected').should("have.text", value);
```
did not.

And yes
```js
cy.get("[data-XXX='XXX']").select(value);
```
also doesn't do anything.

**Comment by smargoli2** [comment]
I have the same issue with clicking a dropdown. It works sometimes but when I put in a click twice it always works.

**Comment by dhorkin** [comment]
I'm also seeing this, intermittently.

Here's example relevant HTML source code:
![image](https://github.com/cypress-io/cypress/assets/59899746/c274c1d0-ca92-40bc-9aae-a799e309c642)

and my code:
```ts
cy.get('[data-test=app_detail_page] [data-test=register_contract]')
    .within(() => {
        cy.get('[data-test=type_selector]')
            .select('sanction_registry');
    })
```

And some partially redacted GUI screenshots:
![image](https://github.com/cypress-io/cypress/assets/59899746/eeee5f8c-7b18-4ffa-8d72-0ecc5294346c)

![image](https://github.com/cypress-io/cypress/assets/59899746/e5c22749-60d2-43ad-9fdb-855d7dfa1b2a)

![image](https://github.com/cypress-io/cypress/assets/59899746/325e19a9-6285-4276-969a-5f4c7f7f2f71)

**Comment by dhorkin** [comment]
Well, this is interesting... I tried overwriting the `select` command as follows (I am aware this has issues like not working if you pass indexes, or issues w/ multi-select):
```ts
Cypress.Commands.overwrite('select', <Subject>(originalSelectFn: Cypress.CommandOriginalFnWithSubject<'select', Subject>, prevSubject: Subject, valueOrTextOrIndex: string | number | Array<string | number>, options?: Partial<Cypress.SelectOptions>) => {
    return originalSelectFn(prevSubject, valueOrTextOrIndex, options)
        .then((subject) => {
            const selected = subject.find(':selected');
            const selectedValues = selected.val();
            const selectedText = selected.text();

            if (!(
                Cypress._.isEqual(selectedValues, valueOrTextOrIndex)
                || Cypress._.isEqual(selectedText, valueOrTextOrIndex)
            )) {
                console.log('RETRYING');
                return originalSelectFn(prevSubject, valueOrTextOrIndex, options);
            }
            console.log('NOT RETRYING');
            return subject;
        });
});
```

It prints "NOT RETRYING" - so it seems like perhaps jquery is saying elements are selected, but the UI hasn't updated accordingly? That would match what is in my screenshots.

it seems like perhaps jquery is saying elements are selected, but the UI hasn't updated accordingly? That would match what is in my screenshots.

Perhaps this is related to https://stackoverflow.com/questions/15195227/jquery-option-select-not-working-in-chrome ?

**Comment by smargoli2** [comment]
I figured out my issue; it was because we load the data asynchronously into the form, so cypress is clicking into the dropdown too quickly, before the options are loaded. I added code to intercept the request and wait until it is loaded.
```
cy.intercept('api/Customers/GetAll').as('getCustomers');
cy.visit('sales/salesOrderForm');
cy.wait(['@getCustomers']);
```

**Comment by CodeMeister362** [comment]
I had the same issue with my query select field and solved it by force selecting twice. TY!

**Comment by gnsoria** [comment]
I started having this problem tonight as well. For me, it seems to only happen when the `select()` call is the first action of the test. If I put assertions in front of it, I need to call it twice. If I put some other action first, I need to call it once.

Calling twice is a problematic solution in my use-case because the field becomes disabled after selecting an option. I have a helper function that is calling `select()` and since that function is unaware of whether it's the first action of the test, if I call once or twice, it will fail in some cases. Because of this, I've added a hard `cy.wait()` before the first-action `select()` calls. I dislike that, but it seems to work reliably.

**Comment by davidzambrana** [comment]
In my case, I only solved it by force selecting twice as well. Waiting or doing other stuff before the first click didn't work.

**Comment by CodeMeister362** [comment]
For my particular instance, I solved it when I realized the cy.select() was having a hard time determining which element to grab because it's React lol the other components are right there too we just can't see them. So, I added name properties on container elements that were indicative of the page I was on and drilled down from there or put the property right on the element itself. 

I hope this helps someone.

**Comment by jrutila** [comment]
I have this same problem. Only having one of these calls
```
      cy.get("select[name='remote']").select("Use zip files", { force: true })
      //cy.get("select[name='remote']").select("Use zip files", { force: true }
```
results the selection not happening and following error in console:
```
cypress_runner.js:3967
Unhandled rejection TypeError: Cannot read properties of null (reading 'get')
at Object.verify (http://localhost:3001/__cypress/runner/cypress_runner.js:113111:35)    
at perform (http://localhost:3001/__cypress/runner/cypress_runner.js:112563:69)
From previous event:    
at Promise.longStackTracesCaptureStackTrace [as _captureStackTrace] (http://localhost:3001/__cypress/runner/cypress_runner.js:3486:19)    
at Promise._then (http://localhost:3001/__cypress/runner/cypress_runner.js:1239:17)    
at Promise.then (http://localhost:3001/__cypress/runner/cypress_runner.js:1132:17)    
at next (http://localhost:3001/__cypress/runner/cypress_runner.js:144490:101)    
at next (http://localhost:3001/__cypress/runner/cypress_runner.js:144455:16)    
at <unknown> (http://localhost:3001/__cypress/runner/cypress_runner.js:144511:16)    
at tryCatcher (http://localhost:3001/__cypress/runner/cypress_runner.js:1807:23)   
 at Promise._settlePromiseFromHandler (http://localhost:3001/__cypress/runner/cypress_runner.js:1519:31)    
at Promise._settlePromise (http://localhost:3001/__cypress/runner/cypress_runner.js:1576:18)    
at Promise._settlePromise0 (http://localhost:3001/__cypress/runner/cypress_runner.js:1621:10)    
at Promise._settlePromises (http://localhost:3001/__cypress/runner/cypress_runner.js:1701:18)    
at <unknown> (http://localhost:3001/__cypress/runner/cypress_runner.js:2392:25)
```

If I add `cy.wait(3000)` before the select, it works!

**Comment by thiagomini** [comment]
Hey guys, I ran into this problem and found a solution for my React + Vite app.

**TL:DR**: Make sure to determine whether you are using a **controlled** or **uncontrolled** [React input](https://react.dev/reference/react-dom/components/input). In my case, I was using a **controlled** input, which means React is responsible for updating its value, and not the DOM. If you are using a controlled input, you **must** include a fake "parent component" in the test that controls the value state of its child.

The issue was relatively simple in my case, but not immediately apparent. Here's the example component:

```tsx
export interface DropdownProps {
  options: {
    value: string;
    label: string;
  }[];
  value?: string;
  onChange?: (value: string) => void;
  name?: string;
}

function Dropdown({
  options,
  value,
  onChange: onSelect,
  name = 'select',
}: DropdownProps) {
  const handleSelectChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    onSelect?.(event.target.value);
  };
  return (
    <div className="flex flex-col space-y-1 max-w-[20%]">
      <label htmlFor="dropdown" className="text-xl font-medium text-gray-700">
        Select
      </label>
      <select
        id="dropdown"
        value={value}
        onChange={handleSelectChange}
        className="cursor-pointer border border-gray-300 rounded-md p-2 text-sm focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-blue-500"
        name={name}
      >
        {options.map((option) => (
          <option key={option.value} value={option.value}>
            {option.label}
          </option>
        ))}
      </select>
    </div>
  );
}

export default Dropdown;
```

And here's how the test looked like:

```tsx
import { useState } from 'react';
import Dropdown from '../../src/components/Dropdown';

describe('Dropdown component', () => {
  it('displays given options and value', () => {
    const options = [
      {
        value: '1',
        label: 'Easy',
      },
      {
        value: '2',
        label: 'Medium',
      },
      {
        value: '3',
        label: 'Hard',
      },
    ];

    cy.mount(<Dropdown options={options} value={'Easy'} name="difficulty" />);

    cy.get('select[name="difficulty"] option:selected')
      .should('be.visible')
      .and('have.text', 'Easy')
      .and('have.value', '1');

    cy.get('select[name="difficulty"').select('Medium');
    cy.get('select').should('have.value', '2');

    cy.get('select[name="difficulty"').select('Hard');
    cy.get('select[name="difficulty"').should('have.value', '3');
  });
});
```

As you can see, I used the `value` in the Dropdown component, but did not pass the `onChange` or handle its state. Then, I changed the test to include a parent "TestComponent", and it works like a charm:

```tsx
import { useState } from 'react';
import Dropdown from '../../src/components/Dropdown';

describe('Dropdown component', () => {
  it('displays given options and value', () => {
    const options = [
      {
        value: '1',
        label: 'Easy',
      },
      {
        value: '2',
        label: 'Medium',
      },
      {
        value: '3',
        label: 'Hard',
      },
    ];

    function TestComponent() {
      // Since the Dropdown component is a controlled input,
      // we need a test component to control the internal state, otherwise the test fails.
      const [value, setValue] = useState('1');
      return (
        <Dropdown
          options={options}
          value={value}
          onChange={setValue}
          name="difficulty"
        />
      );
    }

    cy.mount(<TestComponent />);

    cy.get('select[name="difficulty"] option:selected')
      .should('be.visible')
      .and('have.text', 'Easy')
      .and('have.value', '1');

    cy.get('select[name="difficulty"').select('Medium');
    cy.get('select').should('have.value', '2');

    cy.get('select[name="difficulty"').select('Hard');
    cy.get('select[name="difficulty"').should('have.value', '3');
  });
});
```

I hope it helps someone!

**Comment by marielauredon98** [comment]
It seems that the problem can be bypassed by selecting twice. ✅ 

However, I noticed that `cy.select` works when React has finished its two renders (React.StrictMode). If `cy.select` is executed before the second render, it fails. 

Are you experiencing the same issue? Is there a “cleaner” solution?

**Comment by gnsoria** [comment]
@marielauredon98 A hard wait (i.e. `cy.wait(250)`, give or take a few milliseconds) works reliably for me, but is obviously not ideal testing practice.
